### PR TITLE
feat(Core/DBLayer): replace `char const*` to `std::string_view`

### PR DESCRIPTION
**Pull request by Winfidonarleyan**
## Changes Proposed:
- Replace `char const*` to `std::string_view`
- Fix incorrect data in `creature_template` and `creature_onkill_reputation` 
- Add define `NOMINMAX` for MSVC compiler
- Add template function `Get<>()` for `Field`
- Add template function `SetData` for `PrepareStatment`
- Replace parameters function `PAppend`/`PExecute`/`PQuery` to `Append`/`Execute`/`Query` and add support `fmt::format`

## Issues Addressed:
- Part #10356

## Tests Performed:
- Locally and in my core

## Modules dependence:
- https://github.com/azerothcore/mod-ah-bot/pull/50
- https://github.com/azerothcore/mod-anticheat/pull/44
- https://github.com/azerothcore/Eluna/pull/19
- https://github.com/azerothcore/mod-ip-tracker/pull/6
- https://github.com/azerothcore/mod-transmog/pull/67

## How to Test the Changes:
1. Just startup core and check logs

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by Kitzunu** [comment]
Remind me to review Friday if I havent already

**Review by Kitzunu — APPROVED** [review]

**Comment by FrancescoBorzi** [comment]
you forgot to add the [run-build] label and now master is broken...

File: src/server/database/Database/Field.cpp [hunk 1]
```diff
@@ -28,282 +30,310 @@ Field::Field()
...
-    if (!IsType(DatabaseFieldTypes::Int8))
+    template<typename T>
+    inline bool IsCorrectFieldType(DatabaseFieldTypes type)
     {
-        LogWrongType(__FUNCTION__);
-        return 0;
+        // Int8
+        if constexpr (std::is_same_v<T, bool> || std::is_same_v<T, int8> || std::is_same_v<T, uint8>)
+        {
+            if (type == DatabaseFieldTypes::Int8)
```
Comment (VincentVanclef) [inline src/server/database/Database/Field.cpp:56]: Question, why not just do

```
return type == DatabaseFieldTypes::Int8;
```

Else the function will just run and validate the next 10 IF statements for no reason

**Comment by FrancescoBorzi** [comment]
@Winfidonarleyan macOS build is still broken because of this PR.

Please fix or revert it asap.

```
Undefined symbols for architecture x86_64:
  "std::__1::enable_if<!(std::is_base_of_v<std::__1::basic_string_view<char, std::__1::char_traits<char> >, long>), void>::type PreparedStatementBase::SetValidData<long>(unsigned char, long const&)", referenced from:
      std::__1::enable_if<(std::is_arithmetic_v<long>) || (std::is_same_v<std::__1::vector<unsigned char, std::__1::allocator<unsigned char> >, long>), void>::type PreparedStatementBase::SetData<long>(unsigned char, long) in libgame.a(CalendarMgr.cpp.o)
      std::__1::enable_if<(std::is_arithmetic_v<long>) || (std::is_same_v<std::__1::vector<unsigned char, std::__1::allocator<unsigned char> >, long>), void>::type PreparedStatementBase::SetData<long>(unsigned char, long) in libgame.a(Player.cpp.o)
ld: symbol(s) not found for architecture x86_64
```

File: src/server/database/Database/Field.cpp (hunk 1 above)
Comment (Winfidonarleyan) [reply src/server/database/Database/Field.cpp:56]: Need check db type and c++ type

**Comment by hahalmtdhl** [comment]
How to test:

You can create a new data table in the "world" database or "characters". The data table can be a field or any field, but it must be a "string field"

Query or execute this string through Lua An error is returned:

`> Incorrect value 'qwe' for type 'unsigned int'. Value is raw ? ' false'

> Table name '_ qwe'. Field name 'name'`

------------------------

Only numbers are OK I think any string will be converted

Here are my test examples:

--------------------------------

`DROP TABLE IF EXISTS `_ qwe`;

CREATE TABLE `_ qwe` (

`name` varchar(255) CHARACTER SET utf8mb4 NOT NULL DEFAULT ''

) ENGINE=InnoDB DEFAULT CHARSET=utf8;

INSERT INTO `_ qwe` VALUES ('qwe');`

--------------------------------
After creating the data table, execute the following in Lua:
`local Wquery = CharDBQuery("SELECT * FROM `_qwe` WHERE `name` = "..qwe.." ;")`
Or
`local Wquery = CharDBQuery("SELECT * FROM `_qwe` WHERE `name` = \""..qwe.."\" ;")`
Or
`local Wquery = CharDBQuery("SELECT * FROM `_qwe` WHERE name ;")`

**Comment by Nyeriah** [comment]
Other users have reported the same erroer with text fields (character_settings table) after the database layer rewrite

**Comment by hahalmtdhl** [comment]
> 其他用户在数据库层重写后报告了与文本字段（character_settings表）相同的错误

You may have made the same mistake as me?

**Comment by hahalmtdhl** [comment]
> Other users have reported the same erroer with text fields (character_settings table) after the database layer rewrite

You may have made the same mistake as me?

**Pull request merged**